### ISSUE TITLE
How to configure lfs.<url>.standalonetransferagent to match custom URL scheme?

### ISSUE THREAD

**Issue by buchanae** [git-lfs/git-lfs#3214]
I'm not sure if this is a deficiency in git-lfs, git, or my brain/code. Here's what I'm trying to do:

I'm writing a standalone transfer agent for accessing object stores, such as OpenStack swift. This project is [tanker](https://github.com/buchanae/tanker).

When a user installs tanker, I'd like them to run one command that teaches git-lfs how to use tanker for all their repos. Next, when they clone a repo that contains a `.lfsconfig` with `lfs.url = tanker+swift://mybucket/mypath`, everything should just work.

I thought I could use something like `git config --global lfs.tanker+swift://.standalonetransferagent tanker`, along with the other git-lfs config for `lfs.customtransfer.tanker...`. I figured git-lfs would use `git config --get-urlmatch` to find my custom transfer agent for new repos. However, it seems like `git config --get-urlmatch` doesn't match as I expect when the `<url>` part is just a URL scheme such as `tanker+swift://`. Here's my experiment:

```
RJHB552 lfstest-7 (master)
git config buchanae.tanker://foo.prop1 one

RJHB552 lfstest-7 (master)
git config buchanae.tanker://foo.prop2 two

RJHB552 lfstest-7 (master)
git config buchanae.tanker://.prop3 three

RJHB552 lfstest-7 (master)
git config --get-urlmatch buchanae 'tanker://foo'
buchanae.prop1 one
buchanae.prop2 two

RJHB552 lfstest-7 (master)
git config --get-urlmatch buchanae 'tanker://bar'
```

In summary, I guess I can't match all "tanker+swift://" URLs. Instead, I could match `tanker+swift://mybucket`, although that requires the user to configure each repo independently (possibly).

So, can I install a custom transfer agent globally that matches just the URL scheme? Or, do I need to ask the user to configure each URL domain (bucket in the case of object storage) separately?

Hope that made sense. Thanks in advance?

**Comment by ttaylorr** [comment]
> In summary, I guess I can't match all "tanker+swift://" URLs. Instead,
> I could match `tanker+swift://mybucket`, although that requires the
> user to configure each repo independently (possibly).

I don't think that it's possible to match with the URL configuration
slots.

Since git config --get-urlmatch doesn't match anything for a URL given
as "tanker+swift://", I don't think that introducing behavior
inconsistent with Git's in Git LFS's configuration matching code is the
way to go forward here.

> So, can I install a custom transfer agent globally that matches just
> the URL scheme? Or, do I need to ask the user to configure each URL
> domain (bucket in the case of object storage) separately?

I think that today, the best that we can do is install configuration
variable per-repository, perhaps via a "tanker install" run in each
repository that a callers wishes to use.

But, I think that it would be useful to support this kind of thing based
on URL-scheme alone, though I think that the existing configurations
variable isn't the way to do it.

I'm curious what others might think, too: @bk2204 and @PastelMobileSuit,
could you chime in?

**Comment by buchanae** [comment]
Actually, even per-bucket config doesn't seem to work.

~/.gitconfig
```
<snip>
[filter "lfs"]
	clean = git-lfs clean -- %f
	smudge = git-lfs smudge -- %f
	required = true
	process = git-lfs filter-process
[lfs "swift://buchanan"]
	standalonetransferagent = tanker
[lfs "customtransfer.tanker"]
	path = tanker
	args = transfer
```

.lfsconfig in repo
```
[lfs]
	url = swift://buchanan/tanker/
```

cloning
```
git clone git@github.com:buchanae/lfstest.git lfstest-8
Cloning into 'lfstest-8'...
remote: Counting objects: 61, done.
remote: Compressing objects: 100% (45/45), done.
remote: Total 61 (delta 21), reused 49 (delta 10), pack-reused 0
Receiving objects: 100% (61/61), 6.33 KiB | 6.33 MiB/s, done.
Resolving deltas: 100% (21/21), done.
Downloading data/bat (100 KB)
Error downloading object: data/bat (3deb7f5): Smudge error: Error downloading data/bat (3deb7f554567e14503e596979e06a43e0dbcd7d4def8c290a1c904870abbe365): batch request: missing protocol: "swift://buchanan/tanker/"

Errors logged to /Users/buchanae/projects/lfstest-2/lfstest-8/.git/lfs/logs/20180830T174811.847507094.log
Use `git lfs logs last` to view the log.
error: external filter 'git-lfs filter-process' failed
fatal: data/bat: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'

```

**Comment by buchanae** [comment]
> I think that today, the best that we can do is install configuration
variable per-repository

Ya, I think I'm headed in that direction. Slightly unfortunate, git-lfs installs to the global config, so when a user clone a repo for the first time, they'll get errors. Next, they can go into the repo and install tanker, and things should start working.

Alternatively, I could ask people to use `git lfs install --local`, or just tell people to ignore the error. Not the end of the world, but not great either.

**Comment by buchanae** [comment]
What if you allowed `lfs.standalonetransferagent` in the `.lfsconfig` file? That way, at least git-lfs could look for global config for tanker, and if it didn't exist, it could also communicate that it was missing a specific transfer agent by its name.
